File: src/main/java/net/minestom/server/utils/chunk/ChunkCache.java
package net.minestom.server.utils.chunk;

import net.minestom.server.coordinate.CoordConversion;
import net.minestom.server.instance.Chunk;
import net.minestom.server.instance.Instance;
import net.minestom.server.instance.block.Block;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.UnknownNullability;

@ApiStatus.Internal
public final class ChunkCache implements Block.Getter {
    private final Instance instance;
    private @Nullable Chunk chunk;

    private final @Nullable Block defaultBlock;

    public ChunkCache(Instance instance, @Nullable Chunk chunk,
                      @Nullable Block defaultBlock) {
        this.instance = instance;
        this.chunk = chunk;
        this.defaultBlock = defaultBlock;
    }

    public ChunkCache(Instance instance, Chunk chunk) {
        this(instance, chunk, Block.AIR);
    }

    @Override
    public @UnknownNullability Block getBlock(int x, int y, int z, Condition condition) {
        Chunk chunk = this.chunk;
        final int chunkX = CoordConversion.globalToChunk(x);
        final int chunkZ = CoordConversion.globalToChunk(z);
        if (chunk == null || !chunk.isLoaded() ||
                chunk.getChunkX() != chunkX || chunk.getChunkZ() != chunkZ) {
            this.chunk = chunk = this.instance.getChunk(chunkX, chunkZ);
        }
        if (chunk != null) {
            chunk.lockReadLock();
            try {
                return chunk.getBlock(x, y, z, condition);
            } finally {
                chunk.unlockReadLock();
            }
        } else return defaultBlock;
    }
}

File: src/main/java/net/minestom/server/utils/chunk/ChunkCache.kt
package net.minestom.server.utils.chunk

import net.minestom.server.utils.chunk.ChunkUtils
import net.minestom.server.utils.chunk.ChunkCallback
import java.util.concurrent.CompletableFuture
import java.lang.Void
import java.util.concurrent.atomic.AtomicInteger
import net.minestom.server.utils.callback.OptionalCallback
import java.lang.IllegalArgumentException
import net.minestom.server.utils.function.IntegerBiConsumer
import net.minestom.server.coordinate.Vec
import net.minestom.server.instance.Chunk
import net.minestom.server.instance.Instance
import net.minestom.server.instance.block.Block
import org.jetbrains.annotations.ApiStatus
import org.jetbrains.annotations.UnknownNullability

@ApiStatus.Internal
class ChunkCache @JvmOverloads constructor(
    private val instance: Instance, private var chunk: Chunk?,
    private val defaultBlock: Block = Block.AIR
) : Block.Getter {
    override fun getBlock(x: Int, y: Int, z: Int, condition: Block.Getter.Condition): @UnknownNullability Block? {
        var chunk = chunk
        val chunkX = ChunkUtils.getChunkCoordinate(x)
        val chunkZ = ChunkUtils.getChunkCoordinate(z)
        if (chunk == null || chunk.chunkX != chunkX || chunk.chunkZ != chunkZ) {
            chunk = instance.getChunk(chunkX, chunkZ)
            this.chunk = chunk
        }
        return if (chunk != null) {
            synchronized(chunk) { return chunk.getBlock(x, y, z, condition) }
        } else defaultBlock
    }
}